EventGenerator/src/GammaConvGenerator_module.cc

```diff
@@ -0,0 +1,153 @@
+#include "Offline/GlobalConstantsService/inc/GlobalConstantsHandle.hh"
+#include "Offline/GlobalConstantsService/inc/ParticleDataList.hh"
+#include "Offline/GlobalConstantsService/inc/PhysicsParams.hh"
+#include "Offline/Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Offline/DataProducts/inc/PDGCode.hh"
+#include "Offline/MCDataProducts/inc/StageParticle.hh"
+#include "Offline/Mu2eUtilities/inc/simParticleList.hh"
+
+#include <TTree.h>
+namespace mu2e {
+
+  //================================================================
+  class GammaConvGenerator : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<art::InputTag> inputSimParticles{Name("inputSimParticles"),Comment("A SimParticleCollection with input converted photons.")};
+      fhicl::Atom<unsigned> verbosity{Name("verbosity")};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit GammaConvGenerator(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    double particleMass_;
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+
+    unsigned verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    ProcessCode process;
+  };
+
+  //================================================================
+  GammaConvGenerator::GammaConvGenerator(const Parameters& conf)
+    : EDProducer{conf}
+    , particleMass_(GlobalConstantsHandle<ParticleDataList>()->particle(static_cast<PDGCode::type>(PDGCode::e_minus)).mass())
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+  {
+    produces<mu2e::StageParticleCollection>();
+    process = ProcessCode::mu2eGammaConversion;
+  }
+
+  //================================================================
+  void GammaConvGenerator::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
+
+    if(gammas.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"GammaConvGenerator::produce(): no suitable converted gamma in the input SimParticleCollection\n";
+    }
+
+    // Take the converted gamma as SimParticle
+    const auto gammapart = gammas.at(eng_.operator unsigned int() % gammas.size());
+    const auto p4_gamma = gammapart->endMomentum();
+    const auto p_gamma   = p4_gamma.vect().mag();
+    const auto t_gammastop = gammapart->endGlobalTime();
+
+    // Start generating the e-:
+    // extract flat momentum between 0 and p_gamma
+    double p_eminus = randFlat_.fire(0., p_gamma);
```

Comment:
Since this module produces a flat momentum distribution, while the physical distribution isn't perfectly flat, I suggest to rename this module GammaConvFlat

Reply:
changed, I am in contact with Michael Mackenzie to prepare a more accurate generator taking SImParticle collection as input